src/dawn/Optimizer/Interval.cpp

```diff
@@ -175,37 +176,84 @@ std::vector<Interval> Interval::computePartition(std::vector<Interval> const& in
+
+        *curLowIt = splitLowInterval;
+        *curTopIt = splitHighInterval;
+
+        change = true;
+      } else if(curLowInterval.contains(curTopInterval) &&
+                (curLowInterval.upperBound() == curTopInterval.upperBound())) {
+        Interval splitLowInterval(curLowInterval.lowerLevel(), curTopInterval.lowerLevel(),
+                                  curLowInterval.lowerOffset(), curTopInterval.lowerOffset() - 1);
+        Interval splitHighInterval(curTopInterval.lowerLevel(), curTopInterval.upperLevel(),
+                                   curTopInterval.lowerOffset(), curTopInterval.upperOffset());
+
+        *curLowIt = splitLowInterval;
+        *curTopIt = splitHighInterval;
+
+        change = true;
+      }
+      // if the lower one competely covers the higher one, we need three intervals: the lower one,
+      // the intersection and the top interval
+      else if(curLowInterval.contains(curTopInterval)) {
+        Interval splitLowInterval(curLowInterval.lowerLevel(), curTopInterval.lowerLevel(),
+                                  curLowInterval.lowerOffset(), curTopInterval.lowerOffset() - 1);
+        Interval splitMidInterval(curTopInterval.lowerLevel(), curTopInterval.upperLevel(),
+                                  curTopInterval.lowerOffset(), curTopInterval.upperOffset());
+        Interval splitHighInterval(curTopInterval.upperLevel(), curLowInterval.upperLevel(),
+                                   curTopInterval.upperOffset() + 1, curLowInterval.upperOffset());
+
+        *curLowIt = splitLowInterval;
+        *curTopIt = splitMidInterval;
+        newIntervals.insert(curTopIt, splitHighInterval);
+
+        change = true;
       }
       // otherwise we will generate three intervals: the intersection and the two disjoint intervals
       // of the XOR
       else if(curLowInterval.overlaps(curTopInterval)) {
-        Interval splitLowInterval(curLowInterval.lowerBound(), curTopInterval.lowerBound());
-        Interval splitMidInterval(curTopInterval.lowerBound() + 1, curLowInterval.upperBound());
-        Interval splitHighInterval(curLowInterval.upperBound() + 1, curTopInterval.upperBound());
+        Interval splitLowInterval(curLowInterval.lowerLevel(), curTopInterval.lowerLevel(),
+                                  curLowInterval.lowerOffset(), curTopInterval.lowerOffset() - 1);
+        Interval splitMidInterval(curTopInterval.lowerLevel(), curLowInterval.upperLevel(),
+                                  curTopInterval.lowerOffset(), curLowInterval.upperOffset());
+        Interval splitHighInterval(curLowInterval.upperLevel(), curTopInterval.upperLevel(),
+                                   curLowInterval.upperOffset() + 1, curTopInterval.upperOffset());
 
         *curLowIt = splitLowInterval;
         *curTopIt = splitMidInterval;
         newIntervals.insert(curTopIt, splitHighInterval);
+        change = true;
+      }
+
+      if(change) {
```

Comment:
can you document a bit the algorithm, when is change triggered and what is supposed to do with the intervals?

Reply:
it has a description of what the algorithm is supposed to do in the header. I can update the final statement if you think it would help readability?

Reply:
The header should contain what, given an input , the algorithm compute and returns. 
But what I meant here, is to explain a bit what we are doing in the implementation of the algorithm, so that we can understand in the future how the algorithm works. Step, by step 

See 
 https://github.com/MeteoSwiss-APN/dawn/pull/65/files/fd44ef6844633c4e245db5e38f61047d39bb0683#diff-5ff44e30541a7c4ed622483d82ff9ec2R177
and 
https://github.com/MeteoSwiss-APN/dawn/pull/65/files/fd44ef6844633c4e245db5e38f61047d39bb0683#diff-5ff44e30541a7c4ed622483d82ff9ec2R212

since you modified the algorithm and added an `if(change) ` I was asking to add a comment line 
explaining what the algorithm does in case change variable is set, here
https://github.com/MeteoSwiss-APN/dawn/pull/65/files/fd44ef6844633c4e245db5e38f61047d39bb0683#diff-5ff44e30541a7c4ed622483d82ff9ec2R244